### PR TITLE
Brotli v0.5

### PR DESCRIPTION
**Pull request by addaleax**
* Update brotli to the current `HEAD`
* Switch to using the streaming API in the decoder
* Use custom memory allocators that tell V8 about memory allocations (This should take care of #3)
* Update the `large buffer` test file with the current output of `brotli/bin/bro` (not sure why that changed?)
* Implement the single-buffer methods on top of the streaming API instead of their own C++ code

Let me know what you think!

### CODE REVIEW

**Comment by MayhemYDG** [comment]
I'm about to start reviewing this, but I wanted to ask first: is it safe to update brotli to an unfinished version? The latest released version is `0.4.0`, `0.5.0` is still in development.
https://github.com/google/brotli/releases

File: index.js [hunk 1]
```diff
@@ -25,7 +25,12 @@ function compress(input, params, cb) {
     process.nextTick(cb, new Error('Second argument is not a function.'));
     return;
   }
-  encode.compressAsync(input, params, cb);
+  var stream = new TransformStreamEncode(params);
+  var chunks = [];
+  stream.on('error', cb);
+  stream.on('data', function(c) { chunks.push(c); });
+  stream.on('end', function() { cb(null, Buffer.concat(chunks)); });
```
Comment (MayhemYDG) [inline index.js]: According to the [docs](https://nodejs.org/api/buffer.html#buffer_class_method_buffer_concat_list_totallength):

>If totalLength is not provided, it is calculated from the Buffers in the list. This, however, adds an additional loop to the function, so it is faster to provide the length explicitly.

I haven't benchmarked myself but I assume it'd be wise to count the length every time we push a new chunk.
This also applies to the following streams obviously.

File: index.js [hunk 2]
```diff
@@ -78,14 +98,15 @@ TransformStreamEncode.prototype._flush = function(done) {
       return done(err);
     }
     if (output) {
-      that.push(output);
+      for (var i = 0; i < output.length; i++)
+        that.push(output[i]);
```
Comment (MayhemYDG) [inline index.js]: Braces please.
I should add an ESLint config someday.

File: index.js [hunk 3]
```diff
@@ -153,12 +178,13 @@ TransformStreamDecode.prototype._flush = function(done) {
...
     }
     done();
-  });
+  }, !this.sync);
 };
 
-function decompressStream() {
-  return new TransformStreamDecode();
+function decompressStream(params) {
+  return new TransformStreamDecode(params);
```
Comment (MayhemYDG) [inline index.js]: Mmh why did you add `params` to `TransformStreamDecode`? I don't see it being used.

File: src/dec/stream_decode.cc [hunk 4]
```diff
@@ -34,20 +35,34 @@ NAN_METHOD(StreamDecode::Transform) {
...
-    (const uint8_t*) node::Buffer::Data(buffer),
-    node::Buffer::Length(buffer),
-    &obj->mem_input);
+  obj->next_in = (const uint8_t*) node::Buffer::Data(buffer);
+  obj->available_in = node::Buffer::Length(buffer);
 
   Nan::Callback *callback = new Nan::Callback(info[1].As<Function>());
-  Nan::AsyncQueueWorker(new StreamDecodeWorker(callback, obj, false));
+  StreamDecodeWorker *worker = new StreamDecodeWorker(callback, obj);
+  if (info[2]->BooleanValue()) {
```
Comment (MayhemYDG) [inline src/dec/stream_decode.cc]: I think it might be wiser to pass the sync parameter to the `StreamDecode` and `StreamEncode` constructors instead, so we don't have to do the JS->C dance on each call.

File: src/enc/get_params.cc [hunk 5]
```diff
@@ -1,15 +1,19 @@
 #include "get_params.h"
 
 using namespace v8;
-using namespace brotli;
 
-BrotliParams getParams(Local<Object> userParams) {
-  BrotliParams params;
+BrotliEncoderParams getParams(Local<Object> userParams) {
+  BrotliEncoderParams params = {
+    BROTLI_MODE_GENERIC,
```
Comment (MayhemYDG) [inline src/enc/get_params.cc]: Shouldn't this be `BROTLI_DEFAULT_MODE` instead?
https://github.com/google/brotli/blob/master/enc/encode.h#L35

File: index.js (hunk 1 above)
Comment (addaleax) [inline index.js]: > I haven't benchmarked myself but I assume it'd be wise to count the length every time we push a new chunk.

I was actually the one who wrote the benchmarking for `Buffer.concat` in Node core, and there’s a small, but existing difference in performance. So yeah, makes sense.

File: index.js (hunk 2 above)
Comment (addaleax) [inline index.js]: yup, done

File: index.js (hunk 3 above)
Comment (addaleax) [inline index.js]: It’s being passed to the `Transform` constructor, and it’s good practice in general to pass the first argument of stream constructors to the superclass, so that the constructors are compatible and users can set e.g. the high water mark of a stream to have more control over memory usage.

**Comment by addaleax** [comment]
> is it safe to update brotli to an unfinished version?

Probably about as safe as using any other unreleased version of a project – If you feel more comfortable, you can of course leave this PR open until v0.5 is an official release.

File: src/enc/get_params.cc (hunk 5 above)
Comment (addaleax) [inline src/enc/get_params.cc]: Yup, changed it.

File: src/dec/stream_decode.cc (hunk 4 above)
Comment (addaleax) [inline src/dec/stream_decode.cc]: Hmm. If you prefer, I’ll do that, but in my personal experience maintaining as much state on the JS side as possible is a good idea, if only because it tends to be easier for debugging.

Comment (MayhemYDG) [inline src/dec/stream_decode.cc]: Whichever performs best really.

Comment (addaleax) [inline src/dec/stream_decode.cc]: I’ll try and benchmark these, but my instinct would say that when compared against the work of (de)compression, it won’t even be noticeable. ;)

File: src/enc/stream_encode_worker.cc [hunk 6]
```diff
@@ -1,25 +1,24 @@
...
 
-  if (res && output_size > 0) {
-    output_buffer = (char*) malloc(output_size);
-    if (output_buffer == NULL) {
-      res = false;
-    } else {
-      memcpy(output_buffer, buffer, output_size);
-    }
+  if (output_size > 0) {
+    uint8_t* output = static_cast<uint8_t*>(obj->alloc.Alloc(output_size));
```
Comment (MayhemYDG) [inline src/enc/stream_encode_worker.cc]: Is there no need to check that the output isn't NULL here?

File: src/common/stream_coder.h [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+#include "allocator.h"
+#include <nan.h>
+
+using namespace v8;
+
+class StreamCoder : public Nan::ObjectWrap {
+  public:
+    const uint8_t* next_in;
+    size_t available_in;
```
Comment (MayhemYDG) [inline src/common/stream_coder.h]: These two are only used for decoding, they're not common.

File: src/common/allocator.h [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+
+struct Allocator {
+  Allocator() : allocated_unreported_memory(0) {}
+
+  int64_t allocated_unreported_memory;
+
+  struct AllocatedBuffer {
+    size_t size;
+    size_t available;
+    /* char data[...]; */
```
Comment (MayhemYDG) [inline src/common/allocator.h]: Is this comment meaningful?

File: src/enc/stream_encode_worker.cc (hunk 6 above)
Comment (addaleax) [inline src/enc/stream_encode_worker.cc]: Right, done.

File: src/common/stream_coder.h (hunk 7 above)
Comment (addaleax) [inline src/common/stream_coder.h]: Done… I don’t know how long the streaming compression API has been around in brotli, but do you think it might make sense to use that for compressing, too? It probably would reduce the code differences between decoding and encoding quite a bit more.

File: src/common/allocator.h (hunk 8 above)
Comment (addaleax) [inline src/common/allocator.h]: Yes, the actual content starts right behind the “official” members of `AllocatedBuffer`.

File: src/common/stream_coder.h (hunk 7 above)
Comment (MayhemYDG) [inline src/common/stream_coder.h]: Mmh I'm not sure I understand what you mean here.

Comment (addaleax) [inline src/common/stream_coder.h]: Sorry (I guess the language barrier says hi…) – there’s a `BrotliEncoderCompressStream()` method declared in `encode.h` that works pretty much like `BrotliDecompressStream()`, so using that might be nice to avoid code duplication.

I’m not sure how that interacts with “filling the blockSize” the way it’s currently done in `index.js`, though.

**Comment by MayhemYDG** [comment]
Lovely PR, thanks a lot!

Allocating looks good, I think we always check for null now.
Do we free properly though? I'm not so good at C++ so it's hard to understand all of this but:
- if we give memory to a buffer, V8 will GC it for us so we shouldn't free.
- if we're forced to free, we need to do a memcpy first.
- if brotli gives us an error and we end up not giving the memory to buffers, we need to free.

**Comment by addaleax** [comment]
> V8 will GC it for us so we shouldn't free

Well, if we pass something from `alloc.Alloc()` to V8/Node, V8 will take care of *notifying* us that the buffer needs to be freed. The actual freeing can’t be done with `free()`, because the pointer would by a bit off (due to the `AllocatedBuffer` header), which is why I pass `Allocator::NodeFree` to the buffer constructor.

> if we're forced to free, we need to do a memcpy first.

Not sure what you mean by “forced to free”.

> if brotli gives us an error and we end up not giving the memory to buffers, we need to free.

All chunks that we receive from brotli that end up in `pending_output` and are not converted to `Buffer`s will be freed by the `StreamCoder` destructor, if that’s what you mean.

Also, fwiw, the tests pass without errors when using Valgrind, so that’s a good sign everything is correct.

File: src/common/stream_coder.h (hunk 7 above)
Comment (MayhemYDG) [inline src/common/stream_coder.h]: Back from lunch!

Well we should strive to for what's best of course.
I gave a quick look at `BrotliEncoderCompressStream()` and it seems to handle filling the block for us, so we should be able to remove that part of our code.
https://github.com/google/brotli/blob/master/enc/encode.c#L1510-L1519

**Comment by MayhemYDG** [comment]
>Not sure what you mean by “forced to free”.

https://github.com/MayhemYDG/iltorb/blob/master/src/enc/stream_encode_worker.cc#L12-L24
IIRC it was something like the memory was freed once the worker was destroyed and that ended up crashing node, or something like that. My memory is a bit fuzzy, it's been a while. Could be totally wrong too.

I trust you to be more competent that me at this so it's okay.

File: src/common/stream_coder.h (hunk 7 above)
Comment (addaleax) [inline src/common/stream_coder.h]: Cool, looks like that would be usable, yep. Might be nice to do that separately, though, there are already enough changes in this PR. :)

**Comment by addaleax** [comment]
That `memcpy` bit is because the buffer points to brotli’s internal memory, which is freed by `BrotliEncoderDestroyInstance()`, so it’s important that we don’t do any memory management with it.

I was a bit confused by that, too, and I’m not sure what the reason for this design choice in brotli might have been.

**Comment by MayhemYDG** [comment]
LGTM, let me know if this is ready to be merged.

**Comment by addaleax** [comment]
I should have squashed everything that doesn’t pass the tests on its own now and updated to the latest brotli HEAD again – if you want to hold off merging until a release, feel free to do that. I’m off for the day, have a nice evening!

**Comment by MayhemYDG** [comment]
It looks like the changes past the 0.4.0 release are not relevant to us, so it should be fine. I'm just worried to ship a version that compresses differently from a version that could be found in a package manager like homebrew for example.
Thanks again, I'll merge once AppVeyor is done testing and release to npm later.
Have a great evening!

**Comment by MayhemYDG** [comment]
And published.

I should note that decompressing a large buffer became quite slow now.
```
      ✓ should decompress to a large buffer (11607ms)
```
If I increase the buffer size from `16384` to `65536` I get better results:
```
      ✓ should decompress to a large buffer (3104ms)
```
What do you think?

**Comment by addaleax** [comment]
Oh, yeah, feel free to change that – I didn’t benchmark it a lot, just went with the chunk size Node’s `zlib` uses by default.
